sql: enable exponential backoff in read committed stmt retry loop

michae2 opened the pull request:
Increase the defaults of `max_retries_for_read_committed` and `initial_backoff_of_retries_for_read_committed`.

See individual commits for details and release notes.

Informs: #145377

Release note: None

michae2 committed:
sql: increase default max_retries_for_read_committed to 100

In #107044 we added a retry loop for individual statements executing
under Read Committed isolation. We chose 10 as the default retry limit,
but testing has shown that this is too low for high-contention
workloads. (By comparison, automatic transaction retries in conn_fsm are
unlimited.) Increase the default limit to 100.

Informs: #145377

Release note (sql change): This change increases the default value for
the `max_retries_for_read_committed` session variable from 10 to 100.
Testing has shown that some high-contention workloads running under Read
Committed isolation benefit from more statement retries.

michae2 commented:
First three commits are from #146860 and can be ignored. I'm only planning to backport this PR to 25.2.

stevendanna commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -2749,6 +2769,15 @@ func (ex *connExecutor) dispatchReadCommittedStmtToExecutionEngine(
 			ex.sessionTracing.TraceRetryInformation(
 				ctx, "statement", p.autoRetryStmtCounter, p.autoRetryStmtReason,
 			)
+			// Step both the sequence number and the external read timestamp so that
+			// we can see the results of the conflicting transactions that caused us
+			// to fail.
```
What do we mean by the word `external` here?

Have you definitely confirmed that without the Step the next statement happens at the same read timestamp? I ask because over in the txn coord sender, when we hit a retriable error we bump the txn coord senders read timestamp to the earliest write timestamp that the write should have succeed at.  So I'm wondering if (1) we have a bug or I haven't follow that code correctly or (2) if the main impact of this is pushing our timestamp a bit further into the future making it more likely we succeed.

yuzefovich replied:
There are a few mentions of "external" in #108735.

michae2 replied:
Ah, I think I got "external" from this comment, meaning "in relationship to other transactions": https://github.com/cockroachdb/cockroach/blame/680f1757c59f11f69fb04e5fa004cd819800bcf3/pkg/sql/conn_executor_exec.go#L2563-L2574

But confusingly, in the PR Yahor linked, and in other comments in the same file (such as https://github.com/cockroachdb/cockroach/blame/680f1757c59f11f69fb04e5fa004cd819800bcf3/pkg/sql/conn_executor_exec.go#L2046-L2052) "external" actually means "not internal executor"... so there are two different meanings...

I'm going to remove "external" from my comment and from Nathan's comment, since it's not clear this first meaning is used anywhere else.

stevendanna commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -2738,8 +2747,19 @@ func (ex *connExecutor) dispatchReadCommittedStmtToExecutionEngine(
+	initialBackoff := ex.sessionData().InitialBackoffOfRetriesForReadCommitted
+	useBackoff := initialBackoff > 0
+	opts := retry.Options{
+		InitialBackoff:      initialBackoff,
```
I wonder, if we set the initial backoff to 0, does that mean no backoff? If so, perhaps we could use the MaxAttemps configuration item and simplify the loop conditions a bit by always using r.Next()

yuzefovich reviewed (approved):
Last 2 commits <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but curious about backporting this way to 25.2 which already had 25.2.0 and 25.2.1 released - what's the rationale for doing that?

> Reviewed 1 of 1 files at r1, 3 of 5 files at r3, 4 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147869#-:-OS5jJ-D7GYsD0MklgEO:b6jqs2i)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2, @rafiss, and @stevendanna)



<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
Thanks for working to get to the bottom of this!

michae2 committed:
sql: set default initial_retry_backoff_for_read_committed to 2ms

Testing was performed with a 3-node cluster running the following
transaction repeatedly from 32 connections, all trying to lock the same
row:

```sql
BEGIN;
SELECT 1;
SELECT k, v FROM kv WHERE k IN (0) FOR UPDATE;
-- 10 ms pause
UPSERT INTO kv VALUES (0, <random byte>);
COMMIT;
```

This workload was produced using the following commands:

```
cockroach workload init kv \
  --sequential --cycle-length 1 --insert-count 1 \
  --data-loader INSERT --user demo --db defaultdb

cockroach workload run kv \
  --sequential --cycle-length 1 --write-seq S0 \
  --del-percent 0 --read-percent 0 --span-percent 0 \
  --isolation-level read_committed \
  --sel1-writes --sfu-writes --sfu-wait-delay 10ms \
  --user demo --db defaultdb --concurrency 32
```

For this workload, setting `initial_retry_backoff_for_read_committed`
below 500μs caused more than 100 retries, and above 10ms caused
excessively high PMax query latency. 1ms, 2ms, 4ms all performed well.

We pick 2ms as this should be just at or below the typical execution
time of most workloads. 2ms might be too low for longer-running
statements (though still an improvement over 0ms, the current default).

Release note (sql change): Set the default value of session variable
`initial_retry_backoff_for_read_committed` to 2ms. Testing has shown
that some high-contention workloads running under Read Committed
isolation benefit from exponential backoff. 2ms might be too quick of an
initial backoff for longer-running statements, but setting this much
higher than the normal duration of execution will cause excessive delay.

michae2 reviewed (commented):
You're welcome!

> curious about backporting this way to 25.2 which already had 25.2.0 and 25.2.1 released - what's the rationale for doing that?

I was just overeager to get this into customer hands. I've removed it.

bors r=yuzefovich,stevendanna

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147869#-:-OSSa_uT68qiaKIL2nrW:b-eofqx8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss, @stevendanna, and @yuzefovich)



<!-- Sent from Reviewable.io -->
